bookkeeper-server/src/main/java/org/apache/bookkeeper/client/LedgerHandle.java

```diff
@@ -2175,14 +2175,27 @@ void handleBookieFailure(final Map<Integer, BookieId> failedBookies) {
             if (changingEnsemble) {
                 delayedWriteFailedBookies.putAll(failedBookies);
             } else {
-                changingEnsemble = true;
-                triggerLoop = true;
-
                 toReplace = new HashMap<>(delayedWriteFailedBookies);
                 delayedWriteFailedBookies.clear();
                 toReplace.putAll(failedBookies);
 
                 origEnsemble = getCurrentEnsemble();
+
+                for (Map.Entry<Integer, BookieId> entry : toReplace.entrySet()) {
+                    Integer bookieIndex = entry.getKey();
+                    BookieId addr = entry.getValue();
+                    if (origEnsemble.get(bookieIndex).equals(addr)) {
+                        changingEnsemble = true;
```

Comment:
We need to set `changingEnsemble = true` when entering the else code block to avoid a potential race condition. If all the bookies in the toReplace set do not exist in the current ensemble, set changingEnsemble back to false.

What's more, please add a comment to explain why we need this logic. thanks.